tcmur_cmd_handler.c

```diff
@@ -443,6 +443,127 @@ static int handle_flush(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+	 * that the copy manager allows in a single EXTENDED COPY
+	 * target descriptor list.
+	 */
+	val16 = htobe16(RCR_OP_MAX_TARGET_DESC_COUNT);
+	memcpy(&buf[8], &val16, 2);
+
+	/*
+	 * MAXIMUM SEGMENT COUNT: the max number of segment descriptors
+	 * that the copy manager allows in a single EXTENDED COPY
+	 * segment descriptor list.
+	 */
+	val16 = htobe16(RCR_OP_MAX_SEGMENT_DESC_COUNT);
+	memcpy(&buf[10], &val16, 2);
+
+	/*
+	 * MAXIMUM DESCRIPTOR LIST LENGTH: the max length, in bytes,
+	 * of the target descriptor list and segment descriptor list.
+	 */
+	val32 = htobe32(RCR_OP_MAX_DESC_LIST_LEN);
+	memcpy(&buf[12], &val32, 4);
+
+	/*
+	 * MAXIMUM SEGMENT LENGTH: the length, in bytes, of the largest
+	 * amount of data that the copy manager supports writing via a
+	 * single segment.
+	 */
+	val32 = htobe32(RCR_OP_MAX_SEGMENT_LEN);
+	memcpy(&buf[16], &val32, 4);
+
+	/*
+	 * MAXIMUM CONCURRENT COPIES: the max number of EXTENDED COPY
+	 * commands with the LIST ID USAGE field set to 00b or 10b that
+	 * are supported for concurrent processing by the copy manager.
+	 */
+	val16 = htobe16(RCR_OP_TOTAL_CONCURR_COPIES);
+	memcpy(&buf[34], &val16, 2);
+
+	/*
+	 * MAXIMUM CONCURRENT COPIES: the max number of EXTENDED COPY
+	 * commands with the LIST ID USAGE field set to 00b or 10b that
+	 * are supported for concurrent processing by the copy manager.
+	 */
+	buf[36] = RCR_OP_MAX_CONCURR_COPIES;
+
+	/*
+	 * DATA SEGMENT GRANULARITY: the length of the smallest data
+	 * block that copy manager permits in a non-inline segment
+	 * descriptor. In power of two.
+	 */
+	buf[37] = RCR_OP_DATA_SEG_GRAN_LOG2;
+
+	/*
+	 * INLINE DATA GRANULARITY: the length of the of the smallest
+	 * block of inline data that the copy manager permits being
+	 * written by a segment descriptor containing the 04h descriptor
+	 * type code (see 6.3.7.7). In power of two.
+	 */
+	buf[38] = RCR_OP_INLINE_DATA_GRAN_LOG2;
+
+	/*
+	 * HELD DATA GRANULARITY: the length of the smallest block of
+	 * held data that the copy manager shall transfer to the
+	 * application client in response to a RECEIVE COPY RESULTS
+	 * command with RECEIVE DATA service action (see 6.18.3).
+	 * In power of two.
+	 */
+	buf[39] = RCR_OP_HELD_DATA_GRAN_LOG2;
+
+	/*
+	 * IMPLEMENTED DESCRIPTOR LIST LENGTH: the length, in bytes, of
+	 * the list of implemented descriptor type codes.
+	 */
+	buf[43] = RCR_OP_IMPLE_DES_LIST_LENGTH;
+
+	/*
+	 * The list of implemented descriptor type codes: one byte for
+	 * each segment or target DESCRIPTOR TYPE CODE value (see 6.3.5)
+	 * supported by the copy manager,
+	 */
+	buf[44] = XCOPY_SEG_DESC_TYPE_CODE_B2B; /* block --> block */
+	buf[45] = XCOPY_TARGET_DESC_TYPE_CODE_ID; /* Identification descriptor */
+
+	/* AVAILABLE DATA (n-3)*/
+	val32 = htobe32(42);
+	memcpy(&buf[0], &val32, 4);
+
+	copied = tcmu_memcpy_into_iovec(iovec, iov_cnt, buf, sizeof(buf));
+	if (copied != sizeof(buf)) {
+		return tcmu_set_sense_data(sense, HARDWARE_ERROR,
```

Comment:
This is like inquiry's where if the allocation length is too short you send up to allocation bytes worth of data. The initiator needs to send a buffer that can at least hold the available data field, so if the buffer was too small, it can get that value and resend with a bigger buffer. See 4.2.5.6 in spc 5.